From cf763ede50a8ec2c7e4837fda6674ab164cee7d3 Mon Sep 17 00:00:00 2001
From: Mohan Kumar G <quic_mkumarg@quicinc.com>
Date: Mon, 23 Sep 2024 22:36:05 +0530
Subject: [PATCH] UPSTREAM:PROTOCOL:HOSTAPD: Update correct width in wide
bandwidth channel switch

Currently, the new channel width in wide bandwidth channel switch
subfield does not update based on puncturing bitmap during
channel switch.

As per the draft IEEE P802.11be/D7.0 section 35.15.3, the bandwidth
in wide bandwidth channel switch element should include maximum
bandwidth without any disabled channels. The disabled channels
during channel switch is given by puncturing bitmap.

Update the new channel width with maximum bandwidth which does
not include any punctured channels.

Patch-depdendency: r04-005-hostapd-Update-Wide-Bandwidth-Channel-Switch-element.patch

Patch-work: None

Feature: CHANNEL_SWITCH

Signed-off-by: Mohan Kumar G <quic_mkumarg@quicinc.com>
---
 src/ap/ieee802_11.c | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -8641,31 +8641,26 @@ static u8 * hostapd_eid_wb_channel_switc
 					  u8 chan1, u8 chan2)
 {
 	u8 bw;
+	enum oper_chan_width chan_width = hostapd_get_oper_chwidth(hapd->iface->conf);
 
 	/* bandwidth: 0: 40, 1: 80, 160, 80+80, 4: 320 as per
 	 * IEEE P802.11-REVme/D4.0, 9.4.2.159 and Table 9-314. */
 	switch (hapd->cs_freq_params.bandwidth) {
 	case 40:
 		bw = 0;
+		chan_width = CONF_OPER_CHWIDTH_USE_HT;
 		break;
 	case 80:
 		bw = 1;
+		chan_width = CONF_OPER_CHWIDTH_80MHZ;
 		break;
 	case 160:
 		bw = 1;
-		break;
-	case 320:
-		bw = 4;
-		break;
-	default:
-		/* not valid VHT bandwidth or not in CSA */
-		return eid;
-	}
+		chan_width = CONF_OPER_CHWIDTH_160MHZ;
+
+		if (hapd->cs_freq_params.punct_bitmap)
+			break;
 
-	*eid++ = WLAN_EID_WIDE_BW_CHSWITCH;
-	*eid++ = 3; /* Length of Wide Bandwidth Channel Switch element */
-	*eid++ = bw; /* New Channel Width */
-	if (hapd->cs_freq_params.bandwidth == 160) {
 		/* Update the CCFS0 and CCFS1 values in the element based on
 		 * IEEE P802.11-REVme/D4.0, Table 9-314 */
 
@@ -8679,7 +8674,44 @@ static u8 * hostapd_eid_wb_channel_switc
 			chan1 -= 8;
 		else
 			chan1 += 8;
+		break;
+	case 320:
+		bw = 4;
+		chan_width = CONF_OPER_CHWIDTH_320MHZ;
+		break;
+	default:
+		/* not valid VHT bandwidth or not in CSA */
+		return eid;
 	}
+
+	/* check max bandwidth without any disabled channels */
+	punct_update_legacy_bw(hapd->cs_freq_params.punct_bitmap,
+			       hapd->cs_freq_params.channel, &chan_width,
+			       &chan1, &chan2);
+
+	switch (chan_width) {
+	case CONF_OPER_CHWIDTH_USE_HT:
+		/* Wide Bandwidth Channel Switch element is present only
+		 * when the new channel width is wider than 20 MHz */
+		if (chan1 == hapd->cs_freq_params.channel)
+			return eid;
+		bw = 0;
+		break;
+	case CONF_OPER_CHWIDTH_80MHZ:
+	case CONF_OPER_CHWIDTH_160MHZ:
+		bw = 1;
+		break;
+	case  CONF_OPER_CHWIDTH_320MHZ:
+		bw = 4;
+		break;
+	default:
+		/* Not valid operation channel bandwidth */
+		return eid;
+	}
+
+	*eid++ = WLAN_EID_WIDE_BW_CHSWITCH;
+	*eid++ = 3; /* Length of Wide Bandwidth Channel Switch element */
+	*eid++ = bw; /* New Channel Width */
 	*eid++ = chan1; /* New Channel Center Frequency Segment 0 */
 	*eid++ = chan2; /* New Channel Center Frequency Segment 1 */
 
@@ -8761,7 +8793,7 @@ static u8 * hostapd_eid_bw_indication(st
 u8 * hostapd_eid_chsw_wrapper(struct hostapd_data *hapd, u8 *eid)
 {
 	u8 chan1 = 0, chan2 = 0;
-	u8 *eid_len_offset;
+	u8 *eid_len_offset, *start_pos;
 	int freq1;
 
 	if (!hapd->cs_freq_params.channel ||
@@ -8785,6 +8817,7 @@ u8 * hostapd_eid_chsw_wrapper(struct hos
 	*eid++ = WLAN_EID_CHANNEL_SWITCH_WRAPPER;
 	eid_len_offset = eid++; /* Length of Channel Switch Wrapper element */
 
+	start_pos = eid;
 	eid = hostapd_eid_wb_channel_switch(hapd, eid, chan1, chan2);
 
 #ifdef CONFIG_IEEE80211BE
@@ -8794,6 +8827,9 @@ u8 * hostapd_eid_chsw_wrapper(struct hos
 	}
 #endif /* CONFIG_IEEE80211BE */
 
+	if (eid == start_pos)
+		return eid;
+
 	*eid_len_offset = (eid - eid_len_offset) - 1;
 	return eid;
 }
--- a/src/common/hw_features_common.c
+++ b/src/common/hw_features_common.c
@@ -387,7 +387,7 @@ static void punct_update_legacy_bw_80(u8
 
 	switch (bitmap) {
 	case 0x6:
-		*seg0 = 0;
+		*seg0 = pri_chan;
 		return;
 	case 0x8:
 	case 0x4:
@@ -408,7 +408,7 @@ static void punct_update_legacy_bw_80(u8
 		sec_chan = pri_chan - 4;
 
 	if (bitmap & BIT((sec_chan - first_chan) / 4))
-		*seg0 = 0;
+		*seg0 = pri_chan;
 }
 
 
